homeassistant/components/tradfri/config_flow.py

```diff
@@ -166,6 +167,13 @@ def __init__(self):
             psk=key,
             loop=hass.loop
         )
+
+        def on_hass_stop(event):
+            """Close connection when hass stops."""
+            hass.async_add_job(factory.shutdown())
+
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, on_hass_stop)
```

Comment:
Is this shutting down the same observation as the one created when the config entry is created? If not, why do we need to wait until home assistant stops? If it is the same, we should only need to shut it down once, right?

Reply:
It's any observations that pytradfri holds that have been created since loading, if there's a disconnect and reconnect, that'll be a new observation. It's a global shutdown for any connections pytradfri is holding.

Reply:
So then we only need one event listener?

Reply:
Sorry should have worded that better - any factory instances hold observations, I put this next to factory creation to ensure each factory is shutdown.

Looking at the code below this, you're right - we only need to call it once as the api isn't used after. I'll change later.